frontend/compiled/commitment.go

```diff
@@ -0,0 +1,129 @@
+package compiled
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/consensys/gnark/backend/hint"
+	"github.com/consensys/gnark/frontend"
+	"math/big"
+	"sort"
+)
+
+const Dst = "bsb22-commitment"
+
+type Info struct {
+	Committed              []int // sorted list of id's of committed variables
+	nbPrivateCommitted     int
+	HintID                 hint.ID
+	CommitmentIndex        int
+	CommittedAndCommitment []int // sorted list of id's of committed variables AND the commitment itself
+}
+
+func (i *Info) NbPrivateCommitted() int {
+	return i.nbPrivateCommitted
+}
+
+func (i *Info) NbPublicCommitted() int {
+	return i.NbCommitted() - i.NbPrivateCommitted() // hopefully inlined?
+}
+
+func (i *Info) NbCommitted() int {
+	return len(i.Committed)
+}
+
+func (i *Info) Is() bool {
+	return len(i.Committed) != 0
+}
+
+// Too Java?
+func (i *Info) Initialize(committed []int, nbPublicVariables int, compiler frontend.Compiler) (frontend.Variable, error) {
+
+	sort.Ints(committed)
+	removeRedundancy(&committed)
+	nbPublicCommitted := binarySearch(committed, nbPublicVariables)
+	i.nbPrivateCommitted = len(committed) - nbPublicCommitted
+
+	i.Committed = committed
+
+	var commitment frontend.Variable
+	if hintOut, err := compiler.NewHint(bsb22CommitmentComputePlaceholder, 1, i.GetCommittedVariables()...); err != nil {
+		return nil, err
+	} else {
+		commitment = hintOut[0]
+	}
+
+	i.CommitmentIndex = (commitment.(LinearExpression))[0].WireID()
+
+	i.CommittedAndCommitment = append(committed, i.CommitmentIndex) // TODO: Get rid of this field
+	if i.CommitmentIndex <= committed[len(committed)-1] {
+		return nil, fmt.Errorf("commitment variable index smaller than some committed variable indices")
+	}
+
+	i.HintID = hint.UUID(bsb22CommitmentComputePlaceholder)
+
+	return commitment, nil
+}
+
+func bsb22CommitmentComputePlaceholder(*big.Int, []*big.Int, []*big.Int) error {
+	return fmt.Errorf("placeholder function: to be replaced by commitment computation")
```

Comment:
why is this just a placeholder? Would be nice to have a default impl, so that calling R1CS.Solve on a R1CS with commitment info actually works?

Reply:
(that is, without depending on Groth16)

Reply:
The problem is, this function computes the commitment, so it requires the Commitment Key, created during setup. But this hint is created in compilation time, so the Commitment Key is not available.

The only way out of this that I can think of is to instead randomize the Commitment Key during compilation. But that would break modularity by making setup-type tasks bleed into compilation, and it would also bring a big code-generated curve switch-case into the compilation codebase, which as I understand currently doesn't have any.

Reply:
well, what I'm saying is that, technically, the frontend should build a constraint system, that's solvable as is, without going to the next step (proof systems).
So why not just put a hint that do a sha256 or something deterministic with the inputs, without a "key"?

Reply:
Sha256 is only possible for public variables to which the verifier has direct answer. For private variables we will need that Pedersen commitment whose basis comes from the verifying key. Unfortunately this spaghetti quality, that weakens the modular frontend/backend separation is inherent to the scheme, and actually part of the reason why it works.